lib/geoengineer/gps/gps.rb

```diff
@@ -67,6 +82,40 @@ def self.search(nodes, query)
     nodes.select { |n| n.match(project, environment, config, node_type, node_name) }
   end
 
+  # rubocop:disable Metrics/CyclomaticComplexity
+  def self.dereference(nodes, reference)
+    components = reference.match(REFERENCE_SYNTAX)
+    return reference unless components
+
+    nodes = where(nodes, query_from_reference(reference))
+    raise NotFoundError, "for reference: #{reference}" if nodes.empty?
+
+    dereferenced = nodes.map do |node|
+      next node unless components["resource"]
+      node.send("#{components['resource']}_ref", components["attribute"] || 'id')
+    end
+
+    # If the user provided a wildcard query, they expect an array, otherwise they expect an object
+    wildcard_query?(reference) ? dereferenced : dereferenced.first
```

Comment:
I dont like this because it makes it ambiguous about the return type. Either this always return a list or a single object.